iina/MainWindowController.swift

```diff
@@ -1836,7 +1841,70 @@ class MainWindowController: NSWindowController, NSWindowDelegate {
   }
 
   func toggleWindowFullScreen() {
-    window?.toggleFullScreen(self)
+    guard let window = window else { return }
+
+    if (isInFullScreen) {
+      // exit full screen
+      if currentFullScreenIsLegacy {
+        // exit legacy full screen
+        // call delegate
+        windowWillExitFullScreen(Notification(name: Constants.Noti.legacyFullScreen))
+        // stylemask
+        window.styleMask.remove(.borderless)
+        window.styleMask.remove(.fullScreen)
+        // cancel auto hide for menu and dock
+        NSApp.presentationOptions.remove(.autoHideMenuBar)
+        NSApp.presentationOptions.remove(.autoHideDock)
+        // restore window frame ans aspect ratio
+        // firstly resize to a big frame with same aspect ratio for better visual experience
+        let aspectRatio = NSSize(width: player.info.videoWidth ?? AppData.widthWhenNoVideo,
+                                 height: player.info.videoHeight ?? AppData.heightWhenNoVideo)
+        let aspectSize = aspectRatio.shrink(toSize: window.frame.size)
+        let aspectFrame = aspectSize.centeredRect(in: window.frame)
+        window.setFrame(aspectFrame, display: true, animate: false)
+        // then animate to the original frame
+        if let frame = windowFrameBeforeEnteringFullScreen {
+          window.setFrame(frame, display: true, animate: true)
+        } else {
+          player.notifyMainWindowVideoSizeChanged()
+        }
+        window.aspectRatio = aspectRatio
+        windowFrameBeforeEnteringFullScreen = nil
+        // call delegate
+        windowDidExitFullScreen(Notification(name: Constants.Noti.legacyFullScreen))
+      } else {
+        // system full screen
+        window.toggleFullScreen(self)
+      }
+    } else {
+      // enter full screen
+      if Preference.bool(for: .useLegacyFullScreen) {
+        // enter legacy full screen
+        // cache current window frame
+        currentFullScreenIsLegacy = true
+        windowFrameBeforeEnteringFullScreen = window.frame
+        // call delegate
+        windowWillEnterFullScreen(Notification(name: Constants.Noti.legacyFullScreen))
+        // stylemask
+        window.styleMask.insert(.borderless)
+        window.styleMask.insert(.fullScreen)
+        // cancel aspect ratio
+        window.resizeIncrements = NSSize(width: 1, height: 1)
+        // auto hide menubar and dock
+        NSApp.presentationOptions.insert(.autoHideMenuBar)
+        NSApp.presentationOptions.insert(.autoHideDock)
+        // set frame
+        let screen = window.screen ?? NSScreen.main()!
+        window.setFrame(NSRect(origin: NSZeroPoint, size: screen.frame.size), display: true, animate: true)
```

Comment:
Does `.zero` work here? I think `NSZeroPoint` needs to be updated, but you should be able to use `CGPoint.zero` while you wait.